Migrate to rails

tadaskay opened the pull request:
@mjurkus

mjurkus committed:
Initial commit

(cherry picked from commit 88b7851)

mjurkus committed:
Add report types

(cherry picked from commit f4745a6)

mjurkus committed:
Disable forgery protection for API calls

(cherry picked from commit f3d9f07)

mjurkus committed:
Implement simple report creation

(cherry picked from commit dd3175a)

mjurkus committed:
README.md edited online with Bitbucket

(cherry picked from commit 8fa2bdc)

tadaskay committed:
Remove old app

tadaskay committed:
Polish new readme for rails setup

tadaskay committed:
Cleanup gitignore

ViliusKraujutis reviewed (commented):
Ruby. wow 👍 exciting :)

mjurkus committed:
Add reports controller specs for index and show

tadaskay committed:
Restore original license file

tadaskay committed:
Checkin missing wrappers and windows support

tadaskay committed:
Fixup blank failing specs

mjurkus committed:
Add report creation specs

tadaskay committed:
Dockerize new rails app

tadaskay committed:
Run dockerized specs in Travis test build

tadaskay committed:
Update gemfile lock

mjurkus committed:
Add report status

mjurkus committed:
Use report params for creating reports

tadaskay commented on app/models/report.rb:
```diff
@@ -1,4 +1,9 @@
 class Report < ActiveRecord::Base
+  STATUS_CREATED = 10
```
What is the gain of obfuscating status with some codes? Could we simply enumerate the status and store them as-is in database?

mjurkus replied:
Ordering

tadaskay committed:
Support x64-mingw32 platform in bundler

tadaskay committed:
Run db:seed after db:setup

tadaskay commented:
Since I submitted this PR, github does not allow me to 'request changes', so I'll just leave my findings here:

1. When I POST an empty JSON to `/api/reports`, I get text/html returned with `<title>Action Controller: Exception caught</title>` instead of application/json. The same happens when I POST something like `{ "description": "Hello" }`.
2. Create report controller expects a json where the actual report is nested: `{ "report": {...`. Why? If it's a report resource, I should be able to simply create it by sending its representation: `{ "description": "aa", "address": "aa", ...`.
3. Error messages are returned in Lithuanian. Why? As an API consumer, I didn't specify any lang parameter. English should be the default. Also, English strings in locale files are missing.
4. Also need a design how to properly store and return bilingual content, such as `/api/report_types`. Maybe using `lang` query parameter? As of now, Lithuanian language seems to be hard coded into the schema as well.
5. Create report expects `latitude` and `longitude` while error messages refer to fields `lat` and `lng`. Had to look at the source code to find out what was expected.
6. A successful POST should return status 201 Created with a location header containing URI for the newly created resource.
7. Lat and lng are not validated. The latitude must be a number between -90 and 90 and the longitude between -180 and 180.

This is just a quick look. Let's follow up on these first, and then review more thoroughly.

tadaskay committed:
Update bundler to 1.14.6

mjurkus commented:
1. Bug
2. Because of this ` params.require(:report)`
3. If this were an international or multilingual API it would make sense to use english. Now in all our (one) apps we display errors and other messages in lithuanian. Sure we can pass locale with each request. This is easy to change if we ever need it.
4. Yep, wrong decision on my part here to use localized values in database directly. Let's change that in separate PR maybe?
5. Bug
6. Please link to specification how this should be implemented. 
7. Good catch.

tadaskay commented:
Summary of offline discussion: agreed to make API conventional REST, rather then adapting to existing client (and thus resembling the old API).

tadaskay commented:
2 - Since we are requesting report, it makes sense to simply represent it rather than wrapping it. Were there any other related resources, they could be embedded using Hypermedia:
```
Example:
{
  "_embedded": {
    "reporter": {}
  }
}
```
3/4 - Fair enough, we can add on `lang` parameter later on. 
6 - There is no REST specification, but rather standard conventions:
- http://restcookbook.com/HTTP%20Methods/put-vs-post/
- http://www.vinaysahni.com/best-practices-for-a-pragmatic-restful-api#useful-post-responses
- https://www.w3.org/Protocols/rfc2616/rfc2616-sec9.html#sec9.5

tadaskay committed:
Use slackin badge from c4v.herokuapp.com

Shows actual user count

tadaskay committed:
Disable travis installation step

Since tests are executed in a docker image,
local `bundle install` is an extraneous step

tadaskay committed:
Copy Gemfile.lock to the docker image

So that the locked versions are installed

mjurkus committed:
Drop wrapping in report controller and use corretct params for lat, lng

tadaskay commented on app/controllers/api/reports_controller.rb:
```diff
@@ -1,21 +1,21 @@
 
   def index
-    { :entries => Report.all }
+    Report.all
```
Let's leave wrapping for the collection resource (`entries` is fine), because if we return a JSON array, there is no place to add `_links` (e.g. for paging) or `_embedded`

mjurkus committed:
Update report validation

mjurkus committed:
Restore reports listing nesting

tadaskay commented:
Good work @mjurkus 👍
